pkg/test/ginkgo/external.go

```diff
@@ -24,53 +26,135 @@ type serializedTest struct {
 }
 
-// externalTestsForSuite reads tests from external binary, currently only
-// k8s-tests is supported
-func externalTestsForSuite(ctx context.Context) ([]*testCase, error) {
+// externalTestsForSuite reads tests from external binaries
+func externalTestsForSuite(ctx context.Context, externalTests []string) ([]*testCase, error) {
+	var binaries []TestBinary
+	for _, suite := range externalTests {
+		parts := strings.Split(suite, "=")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid suite %q", suite)
+		}
+		testType := parts[0]
+
+		config := strings.Split(parts[1], ":")
+		if len(config) != 2 {
+			return nil, fmt.Errorf("invalid test %q", suite)
+		}
+
+		imageTag := config[0]
+		path := config[1]
+
+		switch testType {
+		case "gingko":
+			extractedBinary, err := extractGinkgoRunnerBinaryFromReleaseImage(imageTag, path)
+			if err != nil {
+				return nil, fmt.Errorf("failed to extract ginkgo binary from %q: %v", path, err)
+			}
+			binaries = append(binaries, extractedBinary)
+		case "gotest":
+			extractedBinaries, err := extractGoTestRunnerBinariesFromReleaseImage(imageTag, path)
+			if err != nil {
+				return nil, fmt.Errorf("failed to extract ginkgo binary from %q: %v", path, err)
+			}
+			for _, extractedBinary := range extractedBinaries {
+				binaries = append(binaries, extractedBinary)
+			}
+		default:
+			return nil, fmt.Errorf("invalid test %q", suite)
+		}
+	}
+
 	var tests []*testCase
+	for _, binary := range binaries {
+		testCases, err := binary.ListTests(ctx)
+		if err != nil {
+			return nil, err
+		}
+		tests = append(tests, testCases...)
+	}
+	return tests, nil
+}
 
-	// TODO: add support for binaries from other images
-	testBinary, err := extractBinaryFromReleaseImage("hyperkube", "/usr/bin/k8s-tests")
+// extractGinkgoRunnerBinaryFromReleaseImage is responsible for resolving the tag from
+// release image and extracting binary, returns path to the binary or error
+func extractGinkgoRunnerBinaryFromReleaseImage(tag, binary string) (GinkgoTestBinary, error) {
+	tmpDir, err := extractPathFromReleaseImage(tag, binary)
 	if err != nil {
-		return nil, fmt.Errorf("unable to extract k8s-tests binary: %w", err)
+		return "", fmt.Errorf("failed extracting %q from %q: %w", binary, tag, err)
 	}
 
-	command := exec.Command(testBinary, "list")
-	testList, err := runWithTimeout(ctx, command, 1*time.Minute)
+	extractedBinary := filepath.Join(tmpDir, filepath.Base(binary))
+	if err := os.Chmod(extractedBinary, 0755); err != nil {
+		return "", fmt.Errorf("failed making the extracted binary executable: %w", err)
+	}
+	return GinkgoTestBinary(extractedBinary), nil
+}
+
+// extractGoTestRunnerBinariesFromReleaseImage is responsible for resolving the tag from
+// release image and extracting go test binaries, returns path to the binary or error.
+// The expectation is that a tree of directories containing test binaries and their package names
+// exists:
+// /tests/
+// ├── a
+// │ ├── b
+// │ ├── b.package.txt
```

Comment:
n.b. the Go module being tested is information that is lost during `go test -c` ... so we need to record it in a side-channel and consume that here.